File: KMeans.py
from creme import cluster
import pickle
import json
import ast

#read configurations
config = json.loads(input())
savePath = config['savePath']
n_clusters = config['n_clusters']
halflife = config['halflife']
sigma = config['sigma']
mu = config['mu']
random_state = config['randomState']

k_means = cluster.KMeans(n_clusters=n_clusters, halflife=halflife, mu=mu, sigma=sigma, random_state=random_state)

output = {}

while True:

	#wait request
	data = input()

	Xi = json.loads(data)

	output["Predict"] = k_means.predict_one(Xi)

	model = k_means.fit_one(Xi)
	pickle.dump(model, open(savePath, 'wb'))

	print(json.dumps(output))

File: KMeans.js
	 
module.exports = function(RED){
	function KMeansNode(config){
		const path = require('path')
		const utils = require('./utils/utils')

		var node = this;

		//set configurations
		node.file = __dirname + '/KMeans.py'
		node.config = {
			savePath: config.savePath || "KMeans.sav",
			n_clusters: parseInt(config.n_clusters) || 2,
			halflife: Number(config.halflife) || 0.5,
			sigma: Number(config.sigma) || 1,
			mu: Number(config.mu) || 0,
			randomState: Number(config.randomState) || 20
		}
		utils.run(RED, node, config)
	}
	RED.nodes.registerType("KMeans", KMeansNode)
}
